File: python/src/azoth/process/reference/pipe.py
"""``process.pipe`` - a line's pressure drop, with its outlet pressure solved.

Spec: ``specs/models/process/pipe.toml``

The Python twin of ``crates/azoth-process/src/models/pipe.rs`` over
``crates/azoth-process/src/kernels/pipe.rs``, written to mirror them.

# Two equations, one per phase

``AdiabaticPipe.calcPressureOut`` branches on the phase's type. A **gas** takes the
compressible ``P1^2 - P2^2`` form - ``Z``, molar mass and temperature, and no density at
all - and anything else takes Darcy-Weisbach. The branch is the phase label, which is
NeqSim's ``PhaseEos.init`` rule and which this library ports for the hydrate dose pair.

# Two densities, which is the quirk

The liquid branch's *velocity* takes ``getPhysicalProperties().getDensity()`` - the volume
translation applied - and its *Reynolds number* takes ``getKinematicViscosity()``, which
divides the same viscosity by the **untranslated** cubic ``M / (Z R T / P)``. Measured on the
captured butane row the two are ``567.33`` and ``603.86`` kg/m3, 6% apart, and a port that
used one density for both is 6% out on ``Re``.

# The viscosity follows the phase type

``getPhysicalProperties()`` dispatches on it: a gas or a hydrocarbon liquid takes
``PFCTViscosityMethodHeavyOil`` - ``eos.viscosity`` - and an **aqueous** phase takes
``WaterPhysicalProperties``, whose correlation is the liquid ``Viscosity`` class
(``eos.aqueous_viscosity``). Water at 300 K is ``8.5510e-4`` Pa s through the second and
``5.3097e-4`` through the first, so taking one for both puts a water line's Reynolds number
``1.61`` out - which is what this row did until that id existed.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from azoth.core.errors import InvalidInputError
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import Phase, PipeResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos import components as _databank
from azoth.eos.mixture import Mixture
from azoth.eos.reference._mixture_state import reduced_parameters
from azoth.eos.reference.aqueous_viscosity import aqueous_viscosity as aqueous_viscosity_solve
from azoth.eos.reference.hydrate_inhibitor_wt import AQUEOUS, GAS, phase_label
from azoth.eos.reference.ph_flash import enthalpy_at
from azoth.eos.reference.pr_molar_volume import pr_molar_volume
from azoth.eos.reference.pt_flash import pt_flash as pt_flash_solve
from azoth.eos.reference.viscosity import viscosity as viscosity_solve

#: The gas constant, as `eos` uses it everywhere.
R = 8.31446261815324

#: `AdiabaticPipe.run`'s own convergence: `1e-2` bar, and `iter < 25`.
PRESSURE_TOLERANCE = 1.0e3
MAX_PASSES = 25


def pipe(
    components: list[str],
    inlet_n: Q,
    inlet_z: list[float],
    inlet_p: Q,
    inlet_t: Q,
    length: Q,
    diameter: Q,
    roughness: Q,
) -> PipeResult:
    """Drop a stream's pressure along a line.

    Args:
        components: the fluid's substances, by name.
        inlet_n: the inlet's molar flow.
        inlet_z: the inlet composition.
        inlet_p: the inlet pressure.
        inlet_t: the inlet temperature, which is also the outlet's.
        length: the pipe's length.
        diameter: the internal diameter.
        roughness: the sand-grain roughness height.

    Returns:
        The outlet's record and the pressure drop the line implies.

    Raises:
        InvalidInputError: where the geometry is not positive.
        OutOfRangeError: for a temperature the spec bounds.

    Example:
        >>> import azoth
        >>> q = azoth.ureg.Quantity
        >>> r = pipe(
        ...     ["methane", "CO2"],
        ...     q(1.0, "mol/s"),
        ...     [0.7, 0.3],
        ...     q(50.0, "bar"),
        ...     q(300.0, "K"),
        ...     length=q(1000.0, "m"),
        ...     diameter=q(0.1, "m"),
        ...     roughness=q(1e-5, "m"),
        ... )
        >>> round(r.pressure_drop.to("Pa").magnitude, 4)
        21.7722
    """
    spec = _spec()
    checks = checks_for(spec)
    warnings: list[Warning] = []

    n = input_to_si(spec, "inlet_n", inlet_n)
    t = input_to_si(spec, "inlet_t", inlet_t)
    p = input_to_si(spec, "inlet_p", inlet_p)
    length_si = input_to_si(spec, "length", length)
    diameter_si = input_to_si(spec, "diameter", diameter)
    roughness_si = input_to_si(spec, "roughness", roughness)

    apply_checks(
        checks.on_input,
        {
            "inlet_t": t,
            "length": length_si,
            "diameter": diameter_si,
            "roughness": roughness_si,
        }.get,
        warnings,
    )

    states = _route(components, n, inlet_z, p, t, length_si, diameter_si, roughness_si)

    return PipeResult(
        outlet_n=from_si(n, "mol/s"),
        outlet_z=tuple(inlet_z),
        outlet_p=from_si(states.p_out, "Pa"),
        outlet_t=from_si(t, "K"),
        outlet_h=from_si(states.h_out, "J/mol"),
        pressure_drop=from_si(p - states.p_out, "Pa"),
        warnings=tuple(warnings),
    )


@dataclass(frozen=True, slots=True)
class PipeStates:
    """The line's intermediates, in the order it computes them, in SI."""

    #: The outlet pressure the loop settled on.
    p_out: float
    #: The outlet molar enthalpy.
    h_out: float
    #: The velocity the drop was computed at, m/s.
    velocity: float
    #: The Reynolds number it was computed at.
    reynolds: float
    #: The Darcy friction factor.
    friction_factor: float
    #: `laminar`, `transition` or `turbulent`.
    regime: str
    #: The passes the outlet pressure took to settle.
    passes: int


@dataclass(frozen=True, slots=True)
class _Pass:
    """One pass's state, before the loop's pressure has been settled."""

    is_gas: bool
    z_factor: float
    molar_mass: float
    temperature: float
    moles: float
    density: float
    velocity: float
    reynolds: float
    friction_factor: float
    regime: str


def _route(
    components: list[str],
    inlet_n: float,
    inlet_z: list[float],
    inlet_p: float,
    inlet_t: float,
    length: float,
    diameter: float,
    roughness: float,
) -> PipeStates:
    """The line's intermediates, in SI, in the order it computes them."""
    for name, value in (
        ("length", length),
        ("diameter", diameter),
        ("roughness", roughness),
    ):
        if not value > 0.0:
            raise InvalidInputError(
                name, f"a pipe's {name} is a positive length, and {value} is not"
            )

    mixture, ideal_gas = _databank.mixture_of(components, eos="pr")
    area = math.pi / 4.0 * diameter * diameter
    relative_roughness = roughness / diameter

    pressure = inlet_p
    passes = 0
    state = _pass(mixture, inlet_n, inlet_z, inlet_t, pressure, area, diameter, relative_roughness)
    while True:
        passes += 1
        nxt = _outlet_pressure(state, inlet_p, length, diameter)
        settled = abs(nxt - pressure) <= PRESSURE_TOLERANCE
        pressure = nxt
        if settled or passes >= MAX_PASSES:
            break
        state = _pass(
            mixture, inlet_n, inlet_z, inlet_t, pressure, area, diameter, relative_roughness
        )

    # The outlet is a flash at the pressure the loop settled on, at the inlet's temperature:
    # `run` ends with a `TPflash` over the system it has been moving.
    h_out, _ = enthalpy_at(mixture, ideal_gas, inlet_t, pressure, inlet_z)

    return PipeStates(
        p_out=pressure,
        h_out=h_out,
        velocity=state.velocity,
        reynolds=state.reynolds,
        friction_factor=state.friction_factor,
        regime=state.regime,
        passes=passes,
    )


def _pass(
    mixture: Mixture,
    inlet_n: float,
    inlet_z: list[float],
    inlet_t: float,
    pressure: float,
    area: float,
    diameter: float,
    relative_roughness: float,
) -> _Pass:
    """Evaluate the line at one pressure."""
    flash = pt_flash_solve(mixture, from_si(inlet_t, "K"), from_si(pressure, "Pa"), inlet_z)
    # Phase 0 of the flashed system: the liquid when there are two, the single phase when
    # there is one. `pt_flash` reports `x` for the liquid root and `y` for the vapour one.
    if flash.phase == Phase.TWO_PHASE:
        z_factor, composition = flash.z_liquid, list(flash.x)
    elif flash.phase == Phase.ALL_VAPOUR:
        z_factor, composition = flash.z_vapour, list(flash.y)
    else:
        z_factor, composition = flash.z_liquid, list(flash.x)

    reduced = reduced_parameters(mixture, inlet_t, pressure)
    label = phase_label(mixture, reduced, composition, z_factor)
    is_gas = label == GAS

    # In kg/mol, as a bare float: the rest of this arithmetic is SI magnitudes, and a
    # quantity that leaked into it would surface as a comparison against a float.
    masses = [
        zi * component.molar_mass.to("kg/mol").magnitude  # type: ignore[union-attr]
        for zi, component in zip(inlet_z, mixture.components, strict=True)
    ]
    molar_mass = sum(masses)
    # **Two densities.** The translated one is what the liquid branch's velocity takes; the
    # cubic one is what `getKinematicViscosity` divides by. See the module doc.
    volume = pr_molar_volume(z_factor, from_si(inlet_t, "K"), from_si(pressure, "Pa")).v
    density = molar_mass / (volume.to("m**3/mol").magnitude - mixture.volume_shift(inlet_z))
    cubic_density = molar_mass / (z_factor * R * inlet_t / pressure)

    # **The viscosity is the phase's, and NeqSim dispatches on the phase type.** A gas or a
    # hydrocarbon liquid takes the PFCT form; an **aqueous** phase takes
    # `WaterPhysicalProperties` and the liquid `Viscosity` correlation. Water at 300 K is
    # `8.5510e-4` Pa s through the second and `5.3097e-4` through the first.
    if label == AQUEOUS:
        mu = (
            aqueous_viscosity_solve(
                mixture, from_si(inlet_t, "K"), from_si(pressure, "Pa"), composition
            )
            .viscosity.to("Pa*s")
            .magnitude
        )
    else:
        mu = (
            viscosity_solve(mixture, from_si(inlet_t, "K"), from_si(pressure, "Pa"), composition)
            .mu.to("Pa*s")
            .magnitude
        )
    kinematic = mu / cubic_density

    if is_gas:
        volume_rate = inlet_n * z_factor * R * inlet_t / pressure
        velocity = volume_rate / area
    else:
        velocity = inlet_n * molar_mass / density / area

    reynolds = velocity * diameter / kinematic
    friction, regime = _friction_factor(reynolds, relative_roughness)

    return _Pass(
        is_gas=is_gas,
        z_factor=z_factor,
        molar_mass=molar_mass,
        temperature=inlet_t,
        moles=inlet_n,
        density=density,
        velocity=velocity,
        reynolds=reynolds,
        friction_factor=friction,
        regime=regime,
    )


def _outlet_pressure(state: _Pass, inlet_pressure: float, length: float, diameter: float) -> float:
    """The outlet pressure a pass implies, in Pa."""
    if state.is_gas:
        # The compressible general flow equation, `P1^2 - P2^2`, exactly as `run` writes it -
        # including that its leading term is the *mass* flow, `4 n M / pi`.
        leading = (4.0 * state.moles * state.molar_mass / math.pi) ** 2
        dp = (
            leading
            * state.friction_factor
            * length
            * state.z_factor
            * R
            / state.molar_mass
            * state.temperature
            / diameter**5
        )
        return math.sqrt(max(0.0, inlet_pressure**2 - dp))
    # No gravity term: the palette declares no elevations, and `run`'s `dp_gravity` is
    # `rho g (z_in - z_out)` - zero for a level line.
    dp = (
        state.friction_factor
        * length
        / diameter
        * state.density
        * state.velocity
        * state.velocity
        / 2.0
    )
    return inlet_pressure - dp


def _friction_factor(reynolds: float, relative_roughness: float) -> tuple[float, str]:
    """The Darcy friction factor and its regime, as `calcWallFrictionFactor` states them."""
    re = abs(reynolds)
    if re < 1e-10:
        return 0.0, "laminar"
    if re < 2300.0:
        return 64.0 / re, "laminar"
    if re < 4000.0:
        laminar = 64.0 / 2300.0
        turbulent = _haaland(4000.0, relative_roughness)
        return laminar + (turbulent - laminar) * (re - 2300.0) / 1700.0, "transition"
    return _haaland(re, relative_roughness), "turbulent"


def _haaland(reynolds: float, relative_roughness: float) -> float:
    """Haaland's explicit approximation to Colebrook, as the class writes it."""
    inner = 6.9 / reynolds + (relative_roughness / 3.7) ** 1.11
    return (1.0 / (-1.8 * math.log10(inner))) ** 2


def _spec() -> dict[str, object]:
    """The generated spec, imported here so the module has no import-time cycle."""
    from azoth._models_gen import model

    return model("process.pipe")

File: crates/azoth-process/src/kernels/pipe.rs
//! `unit_ops.pipe` - a line's pressure drop, with its outlet pressure solved from the
//! geometry rather than stated.

use azoth_core::units::{Length, Pressure, pascals};
use azoth_core::{AzothError, Result};
use azoth_eos::hydrate_inhibitor_wt::PhaseLabel;
use azoth_eos::{aqueous_viscosity, hydrate_inhibitor_wt, pt_flash, viscosity};

use crate::stream::Stream;

/// The outlet's pressure tolerance and pass cap, from `AdiabaticPipe.run`.
///
/// `1e-2` bar - the class's own figure, and `1e-3` in the Pa this crate works in.
const PRESSURE_TOLERANCE: f64 = 1.0e3;
/// `iter < 25`, the class's own cap.
const MAX_PASSES: usize = 25;
/// The Rackett-free gas constant, as `eos` uses it everywhere.
const R: f64 = 8.31446261815324;

/// What a line did: the outlet record, and the interior a hydraulic calculation reads.
pub struct PipeOutcome {
    /// The outlet record.
    pub outlet: Stream,
    /// The velocity the pressure drop was computed at, m/s.
    pub velocity: f64,
    /// The Reynolds number it was computed at.
    pub reynolds: f64,
    /// The Darcy friction factor.
    pub friction_factor: f64,
    /// Which of the three regimes the friction factor came from.
    pub regime: FlowRegime,
    /// The passes the outlet pressure took to settle.
    pub passes: usize,
}

/// The friction factor's regime, as `AdiabaticPipe` names them.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FlowRegime {
    /// Below `Re = 2300`: `64 / Re`.
    Laminar,
    /// The `2300`-`4000` band, where the class **interpolates** between the laminar and
    /// turbulent values rather than refusing.
    Transition,
    /// Above `4000`: the Haaland form.
    Turbulent,
}

impl FlowRegime {
    /// The label `AdiabaticPipe.getFlowRegime` reports.
    pub fn as_str(self) -> &'static str {
        match self {
            FlowRegime::Laminar => "laminar",
            FlowRegime::Transition => "transition",
            FlowRegime::Turbulent => "turbulent",
        }
    }
}

/// Drop a stream's pressure along a line, solving the outlet pressure it implies.
///
/// **The outlet pressure is a function of the line, not an input** - which is what makes a
/// pipe a different shape of unit operation from the two-port machines. `AdiabaticPipe.run`
/// iterates `calcPressureOut()` against the state it is evaluating at until the two agree to
/// `1e-2` bar or twenty-five passes run out, and the *outlet* is a flash at the pressure it
/// settled on, at the inlet's temperature.
///
/// **Two equations, one per phase**, and the branch is the phase's label - NeqSim's own
/// `PhaseEos.init` rule, which `eos` ports for the hydrate dose pair. A **gas** takes the
/// compressible `P1^2 - P2^2` form; anything else takes Darcy-Weisbach. The two differ in
/// more than algebra: the gas form reads the cubic's `Z` and molar mass, while the liquid
/// form deliberately recomputes the velocity from the *physical-properties* density, because
/// "cubic EOS liquid volumes are inaccurate for polar fluids (e.g. water)".
///
/// # Errors
/// [`azoth_core::AzothError::InvalidInput`] if the geometry is not positive, and whatever
/// the flash, the density, the viscosity or the label refuses.
pub fn pipe(
    feed: &Stream,
    length: Length,
    diameter: Length,
    roughness: Length,
) -> Result<PipeOutcome> {
    for (name, value) in [
        ("length", length.value),
        ("diameter", diameter.value),
        ("roughness", roughness.value),
    ] {
        // `NaN` is not a length either, and this says so without clippy reading the test as
        // a negated comparison on a partial order.
        if value.is_nan() || value <= 0.0 {
            return Err(AzothError::invalid_input(
                name,
                format!("a pipe's {name} is a positive length, and {value} is not"),
            ));
        }
    }

    // The ideal-gas model is the Pitzer departure's, which this kernel does not read: the
    // outlet's enthalpy is a `TP` state, and both branches work from the flash's own `Z`.
    let (mixture, _ideal_gas) = feed.mixture()?;
    let relative_roughness = roughness.value / diameter.value;
    let area = std::f64::consts::PI / 4.0 * diameter.value * diameter.value;

    // The inlet's own pressure, which the compressible form's `P1^2` term holds fixed while
    // the state it is evaluated at walks down to the outlet.
    let inlet_pressure = feed.p.value;
    let mut pressure = feed.p.value;
    let mut passes = 0;
    let mut state = State::at(
        &mixture,
        feed,
        pressure,
        area,
        diameter.value,
        relative_roughness,
    )?;

    loop {
        passes += 1;
        let next = state.outlet_pressure(inlet_pressure, length.value, diameter.value);
        let settled = (next - pressure).abs() <= PRESSURE_TOLERANCE;
        pressure = next;
        if settled || passes >= MAX_PASSES {
            break;
        }
        state = State::at(
            &mixture,
            feed,
            pressure,
            area,
            diameter.value,
            relative_roughness,
        )?;
    }

    // The outlet is a flash at the pressure the loop settled on, at the inlet's temperature:
    // `run` ends with a `TPflash` over the system it has been moving, and never sets a
    // temperature.
    Ok(PipeOutcome {
        outlet: Stream::from_pt(
            feed.components.clone(),
            feed.z.clone(),
            feed.n,
            pascals(pressure),
            feed.t,
        )?,
        velocity: state.velocity,
        reynolds: state.reynolds,
        friction_factor: state.friction,
        regime: state.regime,
        passes,
    })
}

/// One pass's state, and the outlet pressure it implies.
struct State {
    /// The velocity the drop was computed at, m/s.
    velocity: f64,
    /// The Reynolds number.
    reynolds: f64,
    /// The Darcy friction factor.
    friction: f64,
    /// Which regime produced it.
    regime: FlowRegime,
    /// Whether the phase took the compressible branch.
    is_gas: bool,
    /// The compressible form's `Z`, molar mass and temperature.
    z_factor: f64,
    molar_mass: f64,
    temperature: f64,
    /// The molar flow the compressible form's leading term carries.
    moles: f64,
    /// The liquid form's density, from the physical properties.
    density: f64,
}

impl State {
    /// Evaluate the line at one pressure.
    #[allow(clippy::too_many_arguments)] // one per quantity the class reads, and there are six
    fn at(
        mixture: &azoth_eos::Mixture,
        feed: &Stream,
        pressure: f64,
        area: f64,
        diameter: f64,
        relative_roughness: f64,
    ) -> Result<State> {
        let p = pascals(pressure);
        let flash = pt_flash(mixture, feed.t, p, &feed.z)?;
        // Phase 0 of the flashed system, which is the liquid when there are two and the
        // single phase when there is one. `pt_flash` reports `x` for the liquid root and `y`
        // for the vapour one, so a gas state's composition is `y`.
        let (z_factor, composition) = match flash.phase {
            azoth_eos::Phase::TwoPhase => (flash.z_liquid, flash.x.clone()),
            azoth_eos::Phase::AllVapour => (flash.z_vapour, flash.y.clone()),
            azoth_eos::Phase::AllLiquid | azoth_eos::Phase::Trivial => {
                (flash.z_liquid, flash.x.clone())
            }
        };

        let reduced = mixture.reduced_parameters(feed.t, p)?;
        let label = hydrate_inhibitor_wt::label(mixture, &reduced, &composition, z_factor)?;
        let is_gas = label == PhaseLabel::Gas;

        // **Two densities, and the class uses them in different places.** The liquid
        // branch's *velocity* takes the physical-properties density, which applies the
        // Peneloux translation - and its *Reynolds number* takes `getKinematicViscosity()`,
        // which divides the same viscosity by the **untranslated cubic** `M/(Z R T / P)`.
        // Measured on the captured butane row, the two differ by 6%: `567.33` against
        // `604.0`, which is the whole of the Reynolds gap between this port and NeqSim
        // before the distinction is made.
        let density = density_at(mixture, feed, p, z_factor)?;
        let cubic_density =
            molar_mass_of(mixture, &feed.z)? / (z_factor * R * feed.t.value / pressure);
        // **The viscosity is the phase's, and NeqSim dispatches on the phase type.** A gas or
        // a hydrocarbon liquid takes `PFCTViscosityMethodHeavyOil` - `eos.viscosity` - and an
        // **aqueous** phase takes `WaterPhysicalProperties`, whose correlation is the liquid
        // `Viscosity` class: `eos.aqueous_viscosity`. Water at 300 K is `8.5510e-4` Pa s
        // through the second and `5.3097e-4` through the first.
        let mu = match label {
            PhaseLabel::Aqueous => {
                aqueous_viscosity::aqueous_viscosity(mixture, feed.t, p, &composition)?
                    .viscosity
                    .value
            }
            PhaseLabel::Gas | PhaseLabel::Oil => {
                viscosity::viscosity(mixture, feed.t, p, &composition)?
                    .mu
                    .value
            }
        };
        let kinematic = mu / cubic_density;

        // The gas branch takes the velocity from the cubic's total volume; the liquid branch
        // recomputes it from the mass flow and the physical-properties density. They are the
        // same number in the limit and not otherwise, which is why both are written out.
        let molar_mass = molar_mass_of(mixture, &feed.z)?;
        let velocity = if is_gas {
            let volume_rate = feed.n * z_factor * R * feed.t.value / pressure;
            volume_rate / area
        } else {
            feed.n * molar_mass / density / area
        };

        let reynolds = velocity * diameter / kinematic;
        let (friction, regime) = friction_factor(reynolds, relative_roughness);

        Ok(State {
            velocity,
            reynolds,
            friction,
            regime,
            is_gas,
            z_factor,
            molar_mass,
            temperature: feed.t.value,
            moles: feed.n,
            density,
        })
    }

    /// The outlet pressure this state implies, in Pa.
    fn outlet_pressure(&self, inlet_pressure: f64, length: f64, diameter: f64) -> f64 {
        if self.is_gas {
            // The compressible general flow equation, `P1^2 - P2^2`, exactly as `run` writes
            // it - including that its leading term is the *mass* flow, `4 n M / pi`.
            let leading = (4.0 * self.moles * self.molar_mass / std::f64::consts::PI).powi(2);
            let dp = leading * self.friction * length * self.z_factor * R / self.molar_mass
                * self.temperature
                / diameter.powi(5);
            // No gravity term: the palette declares no elevations, and `run`'s
            // `dp_gravity` is `rho g (z_in - z_out)` - zero for a level line. A line with
            // a rise is a column the entry cannot describe, which is stated in its spec
            // rather than defaulted.
            (inlet_pressure.powi(2) - dp).max(0.0).sqrt()
        } else {
            let dp =
                self.friction * length / diameter * self.density * self.velocity * self.velocity
                    / 2.0;
            inlet_pressure - dp
        }
    }
}

/// The Darcy friction factor and the regime it came from, as `calcWallFrictionFactor` states
/// them.
///
/// **The transition band is interpolated, not refused.** Between `2300` and `4000` the class
/// linearly blends the laminar value at `2300` with the Haaland value at `4000`, and this
/// reproduces it: where the class draws a straight line between two points, a port that
/// refused the band would be answering a smaller question than the class does.
fn friction_factor(reynolds: f64, relative_roughness: f64) -> (f64, FlowRegime) {
    let re = reynolds.abs();
    if re < 1e-10 {
        return (0.0, FlowRegime::Laminar);
    }
    if re < 2300.0 {
        return (64.0 / re, FlowRegime::Laminar);
    }
    if re < 4000.0 {
        let laminar = 64.0 / 2300.0;
        let turbulent = haaland(4000.0, relative_roughness);
        return (
            laminar + (turbulent - laminar) * (re - 2300.0) / 1700.0,
            FlowRegime::Transition,
        );
    }
    (haaland(re, relative_roughness), FlowRegime::Turbulent)
}

/// Haaland's explicit approximation to Colebrook, as the class writes it.
fn haaland(reynolds: f64, relative_roughness: f64) -> f64 {
    let inner = 6.9 / reynolds + (relative_roughness / 3.7).powf(1.11);
    (1.0 / (-1.8 * inner.log10())).powi(2)
}

/// The physical-properties density at a state, which is the cubic's volume with the volume
/// translation applied.
fn density_at(
    mixture: &azoth_eos::Mixture,
    feed: &Stream,
    p: Pressure,
    z_factor: f64,
) -> Result<f64> {
    let volume =
        azoth_eos::pr_molar_volume(z_factor, feed.t, p)?.v.value - mixture.volume_shift(&feed.z);
    Ok(azoth_eos::pr_mass_density(
        azoth_core::units::kilograms_per_mole(molar_mass_of(mixture, &feed.z)?),
        azoth_core::units::cubic_meters_per_mole(volume),
    )?
    .rho
    .value)
}

/// The mixture's molar mass, kg/mol.
fn molar_mass_of(mixture: &azoth_eos::Mixture, z: &[f64]) -> Result<f64> {
    let mut total = 0.0;
    for (component, zi) in mixture.components().iter().zip(z) {
        let mass = component.molar_mass.ok_or_else(|| {
            AzothError::property_unavailable(
                "a pipe's fluid",
                "molar_mass",
                "a component of this stream carries no molar mass, so the mixture has none",
            )
        })?;
        total += zi * mass;
    }
    Ok(total)
}
